src/som/vmobjects/SObject.java

```diff
@@ -517,4 +517,22 @@ private void setFieldAfterLayoutChange(final SlotDefinition slot,
     StorageLocation location = getLocation(slot);
     location.write(this, value);
   }
+
+  public final boolean isObjectSlotAllocated(final SlotDefinition slot) {
+    StorageLocation loc = objectLayout.getStorageLocation(slot);
+    return loc instanceof ObjectStorageLocation;
+  }
+
+  public final synchronized void ensureSlotAllocatedToAvoidDeadlock(
```

Comment:
Hmmmmmm. So, we still hold a lock when we may trigger a safepoint? Any way around that?

Seems like this is only moving the deadlock?

Reply:
this method is called after the safepoint starts, and nowhere else. the synchronized is for the case that two threads enter the safepoint to update the same object.